### PR TITLE
refactor: improved typings

### LINKED ISSUE #528: Organize and take more advantage of aliases for TypeScript types
**Describe the feature**
What feature are you proposing?
`import type` for importing only for type declarations would be neater as well.

**Priority**
- [ ] High
- [x] Medium
- [ ] Low

**List the benefits of adding such a feature**
What will you gain from this feature?
N/A

**Is this feature request related to a problem?**
Is there a problem with the current implementation (if any), or is there an issue that would be resolved with this feature?
N/A

### PR DESCRIPTION
**Pull request by zapteryx**
### Thank you for your interest in contributing!
Before proceeding, please review the [guidelines for contributing](https://github.com/ZPTXDev/Quaver/blob/master/CONTRIBUTING.md).

- [x] Are you targeting the `next` branch? (left side)
- [x] Did you review the guidelines for contributing?
- [x] Does your code pass linting checks?
- [x] Is your code documented, if applicable? (Check if not applicable)
- [x] Does this PR have a linked issue?

### Scope of change
- [ ] Major change
- [ ] Minor change
- [x] Documentation only

### Type of change
- [ ] Bug fix
- [ ] Feature
- [x] Other

### Priority
- [ ] Critical
- [ ] High
- [ ] Medium
- [x] Low

### Description
Please describe the changes.
improves typings
Closes #528

### CODE REVIEW

**Review by Xayanide — CHANGES_REQUESTED** [review]
You can either
1.) Write that so that every file that has `.types.` on their file names will be not be included and emitted during compilation, because they were included (the type files) during compilation and the for-of loops tried to read them and gave
```js
$ node .
file:///C:/Users/Ava/Documents/GitHub/Quaver/dist/main.js:226    
    if (event.default.once) {
                      ^
TypeError: Cannot read properties of undefined (reading 'once')  
    at file:///C:/Users/Ava/Documents/GitHub/Quaver/dist/main.js:226:23
```
You can do this by adding and using "exclude"`: [""]` to `tsconfig.json`

2.) Rename every type files to use `.d.ts` instead of `.types.ts`, doesn't involve adding an `"exclude"` to `tsconfig.json` since `skipLibCheck` is true, and update the import paths accordingly.

Or other ways better than this to resolve the problem.

**Review by Xayanide — APPROVED** [review]

File: src/lib/util/settings.d.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	database?: DatabaseSettingsObject;
+	lavalink?: LavalinkSettingsObject;
+	features?: FeaturesSettingsObject;
+};
+
+export type ColorsSettingsObject = {
+    success?: string;
+    neutral?: string;
+    warning?: string;
+    error?: string;
```
Comment (Xayanide) [inline src/lib/util/settings.d.ts:17]: How about `ColorResolvable` and not `string` then remove where you used `ColorResolvable` as type assertions.

**Review by Xayanide — CHANGES_REQUESTED** [review]

**Review by Xayanide — CHANGES_REQUESTED** [review]
Use the type `ConnectionInfo` instead provided by `lavaclient` here. Delete the type `LavalinkSettingsObject`.

File: src/lib/util/common.d.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+export type QuaverChannels = TextChannel | VoiceChannel;
+
+export type QuaverClient = Client<boolean> & {
+    music?: Node;
+    commands: Collection<string, unknown>;
+    buttons: Collection<string, unknown>;
+    selectmenus: Collection<string, unknown>;
+    autocomplete: Collection<string, unknown>;
+    modals: Collection<string, unknown>;
```
Comment (Xayanide) [inline src/lib/util/common.d.ts:39]:
```diff
+ import type { ChatInputCommand, Button, SelectMenu, ModalSubmit } from '#src/events/interactionCreate.d.js';

export type QuaverClient = Client<boolean> & {
    music?: Node;
+    commands: Collection<string, ChatInputCommand>;
+    buttons: Collection<string, Button>;
+    selectmenus: Collection<string, SelectMenu>;
+    autocomplete: Collection<string, Autocomplete>;
+    modals: Collection<string, ModalSubmit>;
};
```

**Review by Xayanide — CHANGES_REQUESTED** [review]
```diff
+ import type { ChatInputCommand, Button, SelectMenu, ModalSubmit } from '#src/events/interactionCreate.d.js';

export type QuaverClient = Client<boolean> & {
    music?: Node;
    commands: Collection<string, ChatInputCommand>;
+    buttons: Collection<string, Button>;
+    selectmenus: Collection<string, SelectMenu>;
+    autocomplete: Collection<string, Autocomplete>;
+    modals: Collection<string, ModalSubmit>;
};
```

File: src/lib/ReplyHandler.d.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export type AdditionalBuilderOptions = {
+    ephemeral?: boolean;
+    fetchReply?: boolean;
+    force?: 'reply' | 'edit' | 'update';
+};
```
Comment (Xayanide) [inline src/lib/ReplyHandler.d.ts:5]:
```suggestion
import type { InteractionReplyOptions, InteractionUpdateOptions, WebhookEditMessageOptions } from 'discord.js';

export type AdditionalBuilderOptions = InteractionReplyOptions & WebhookEditMessageOptions & InteractionUpdateOptions;
```

File: src/lib/ReplyHandler.ts [hunk 4]
```diff
@@ -21,7 +23,7 @@ export default class ReplyHandler {
 	 * @param options - Extra data, such as type or components.
 	 * @returns The message that was sent.
 	 */
-	async reply(inputData: string | EmbedBuilder | (string | EmbedBuilder)[], { type = 'neutral', components = null, files = null, ephemeral = false, fetchReply = false, force = null }: { type?: 'success' | 'neutral' | 'warning' | 'error'; components?: ActionRowBuilder<MessageActionRowComponentBuilder>[]; files?: AttachmentBuilder[]; ephemeral?: boolean; fetchReply?: boolean; force?: 'reply' | 'edit' | 'update'; } = {}): Promise<InteractionResponse | Message | boolean> {
+	async reply(inputData: MessageOptionsBuilderInputs, { type = 'neutral', components = null, files = null, ephemeral = false, fetchReply = false, force = null }: MessageOptionsBuilderOptions & AdditionalBuilderOptions = {}): Promise<InteractionResponse | Message | undefined> {
```
Comment (Xayanide) [inline src/lib/ReplyHandler.ts:26]:
```suggestion
	async reply(inputData: MessageOptionsBuilderInputs, { type = 'neutral', components = null, files = null, ephemeral = false, fetchReply = false, force = null }: MessageOptionsBuilderOptions = {}): Promise<InteractionResponse | Message | undefined> {
```

File: src/lib/util/common.d.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    actionRow: ActionRowBuilder;
+};
+
+export type MessageOptionsBuilderInputs = string | EmbedBuilder | (string | EmbedBuilder)[];
+
+export type MessageOptionsBuilderOptions = {
+	type?: 'success' | 'neutral' | 'warning' | 'error';
+	components?: ActionRowBuilder<MessageActionRowComponentBuilder>[];
+	files?: AttachmentBuilder[];
+};
```
Comment (Xayanide) [inline src/lib/util/common.d.ts:27]:
```suggestion
export type MessageOptionsBuilderOptions = {
	type?: 'success' | 'neutral' | 'warning' | 'error';
    force?: 'reply' | 'edit' | 'update';
} & AdditionalBuilderOptions
```

**Review by Xayanide — CHANGES_REQUESTED** [review]
Make use of what have already been made. Not sure whether `type` and `force` belongs to `AdditionalBuilderOptions` or not.

File: src/lib/ReplyHandler.d.ts (hunk 3 above)
Comment (zapteryx) [reply src/lib/ReplyHandler.d.ts:5]: not intended

File: src/lib/ReplyHandler.ts (hunk 4 above)
Comment (zapteryx) [reply src/lib/ReplyHandler.ts:26]: again, not the intention

File: src/lib/util/common.d.ts (hunk 5 above)
Comment (zapteryx) [reply src/lib/util/common.d.ts:27]: additional is not always part of the base

Comment (Xayanide) [reply src/lib/util/common.d.ts:27]: Right, there was `PlayerHandler` so that makes sense.

File: src/main.ts [hunk 6]
```diff
@@ -194,7 +194,7 @@ export async function shuttingDown(eventType: string, err?: Error): Promise<void
 	}
 }
 
-const locales = new Collection();
+const locales = new Collection<string, unknown>();
```
Comment (Xayanide) [inline src/main.ts:197]:
```suggestion
type LocaleProp = {
	[key: string]: Record<string, string | Record<string, string>>
}
const locales = new Collection<string, LocaleProp>();
```

**Review by Xayanide — CHANGES_REQUESTED** [review]

**Review by Xayanide — CHANGES_REQUESTED** [review]
```js
const eventFiles = readdirSync(getAbsoluteFileURL(import.meta.url, ['events'])).filter((file): boolean => file.endsWith('.js') || file.endsWith('.ts'));
for await (const file of eventFiles) {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	const event: { default: { name: string; once: boolean; execute(...args: any[]): void | Promise<void>; } } = await import(getAbsoluteFileURL(import.meta.url, ['events', file]).toString());
	if (event.default.once) {
		bot.once(event.default.name, (...args): void | Promise<void> => event.default.execute(...args));
	}
	else {
		bot.on(event.default.name, (...args): void | Promise<void> => event.default.execute(...args));
	}
}

const musicEventFiles = readdirSync(getAbsoluteFileURL(import.meta.url, ['events', 'music'])).filter((file): boolean => file.endsWith('.js') || file.endsWith('.ts'));
for await (const file of musicEventFiles) {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	const event: { default: { name: any, once: boolean, execute(...args: any[]): void & Promise<void> } } = await import(getAbsoluteFileURL(import.meta.url, ['events', 'music', file]).toString());
	if (event.default.once) {
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		bot.music.once(event.default.name, (...args: any[]): void => event.default.execute(...args));
	}
	else {
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		bot.music.on(event.default.name, (...args: any[]): void => event.default.execute(...args));
	}
}
```

```js
{ name: any, once: boolean, execute(...args: any[]): void & Promise<void> }
```

Make an alias for them accordingly.

**Review by Xayanide — CHANGES_REQUESTED** [review]
Can you also please provide a type for every `try...catch`'s `local error variable` as well.

**Review by Xayanide — CHANGES_REQUESTED** [review]
```diff
type LocalesProps = {
	[key: string]: Record<string, string | Record<string, string | Record<string, string>>>
}

+ const locales = new Collection<string, LocalesProps >();
const localeFolders = readdirSync(getAbsoluteFileURL(import.meta.url, ['..', 'locales']));
for await (const folder of localeFolders) {
	const localeFiles = readdirSync(getAbsoluteFileURL(import.meta.url, ['..', 'locales', folder]));
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	const localeProps: LocalesProps = {};
	for await (const file of localeFiles) {
		const categoryProps = await import(getAbsoluteFileURL(import.meta.url, ['..', 'locales', folder, file]).toString());
		const categoryName = file.split('.')[0].toUpperCase();
		localeProps[categoryName] = categoryProps.default;
	}
	locales.set(folder, localeProps);
}
```

**Comment by zapteryx** [comment]
> Can you also please provide a type for every `try...catch`'s `local error variable` as well.

that's not how that works

**Comment by zapteryx** [comment]
> ```diff
> + type LocaleProps = {
> +	[key: string]: Record<string, string | Record<string, string | Record<string, string>>>
> + }
> 
> + const locales = new Collection<string, LocaleProps>();
> const localeFolders = readdirSync(getAbsoluteFileURL(import.meta.url, ['..', 'locales']));
> for await (const folder of localeFolders) {
> 	const localeFiles = readdirSync(getAbsoluteFileURL(import.meta.url, ['..', 'locales', folder]));
> -	// eslint-disable-next-line @typescript-eslint/no-explicit-any
> +	const localeProps: LocaleProps = {};
> 	for await (const file of localeFiles) {
> 		const categoryProps = await import(getAbsoluteFileURL(import.meta.url, ['..', 'locales', folder, file]).toString());
> 		const categoryName = file.split('.')[0].toUpperCase();
> 		localeProps[categoryName] = categoryProps.default;
> 	}
> 	locales.set(folder, localeProps);
> }
> ```

will not be implementing this

**Review by Xayanide — CHANGES_REQUESTED** [review]

File: src/main.ts [hunk 7]
```diff
@@ -248,14 +252,14 @@ for await (const file of eventFiles) {
 const musicEventFiles = readdirSync(getAbsoluteFileURL(import.meta.url, ['events', 'music'])).filter((file): boolean => file.endsWith('.js') || file.endsWith('.ts'));
 for await (const file of musicEventFiles) {
 	// eslint-disable-next-line @typescript-eslint/no-explicit-any
-	const event: { default: { name: any, once: boolean, execute(...args: any[]): void & Promise<void> } } = await import(getAbsoluteFileURL(import.meta.url, ['events', 'music', file]).toString());
+	const event: { default: QuaverEvent } = await import(getAbsoluteFileURL(import.meta.url, ['events', 'music', file]).toString());
```
Comment (Xayanide) [inline src/main.ts:255]:
```suggestion
	const event: { default: QuaverMusicEvent } = await import(getAbsoluteFileURL(import.meta.url, ['events', 'music', file]).toString());
```

**Review by Xayanide — CHANGES_REQUESTED** [review]

File: src/main.ts (hunk 7 above)
Comment (Xayanide) [inline src/main.ts:255]:
```suggestion
	const event: { default: QuaverMusicEvent } = await import(getAbsoluteFileURL(import.meta.url, ['events', 'music', file]).toString());
```

**Review by Xayanide — CHANGES_REQUESTED** [review]

**Review by Xayanide — APPROVED** [review]

**Pull request merged**